### PR TITLE
ECRecover contract upgrade

### PR DESCRIPTION
**Pull request by YakupIpek**
The stratisproject/StratisBitcoinFullNode#4194 moved here. 

This pr needs to have package from stratisproject/Stratis.SmartContracts#13

This pr has been reviewed actually and should to be complete but it was require to have hard-fork because upgrade needed on Stratis.SmartContracts assemblies. 

Basically `EcRecoverProvider.SignMessage(privateKey, message)` returns a signature that we can recover `PubKey` address and we can do this via `ECRecoverProvider.GetSigner(message, offChainSignature)` in the smart contract code. 

I reviewed&tested it  and i don't have anything to add. 

I am going to remove dev prefix from version(2.0.1-dev) if it is approved.

### CODE REVIEW

**Review by quantumagi — APPROVED** [review]
LGTM

File: src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        /// <summary>
+        /// Retrieves the base58 address of the signer of an ECDSA signature.
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="signature">The ECDSA signature prepended with header information specifying the correct value of recId.</param>
+        /// <returns>The base58 address for the signer of a signature.</returns>
+        public Address GetSigner(byte[] message, byte[] signature)
+        {
+            // TODO: Error handling for incorrect signature format etc.
```
Comment (rowandh) [inline src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs:34]: I think this needs to be resolved before this can be used in production.

**Comment by quantumagi** [comment]
Perhaps include support for `VerifySignatures` into this PR. See https://github.com/stratisproject/Stratis.SmartContracts/blob/e456b8d9641d8357d264c991d7377e9f5b1a9a6c/Stratis.SmartContracts/IEcRecoverProvider.cs.

File: src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs (hunk 1 above)
Comment (YakupIpek) [reply src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs:34]: @rowandh PubKey.RecoverCompact already detects incorrect format and raise an exception. 

There is something else to do in here ?

File: src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+            uint256 hashedUint256 = GetUint256FromMessage(message);
+            PubKey pubKey = PubKey.RecoverCompact(hashedUint256, signature);
+
+            return pubKey.GetAddress(this.network).ToString().ToAddress(this.network);
```
Comment (quantumagi) [inline src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs:39]: Perhaps this conversion can be done without using `this.network`.

Comment (YakupIpek) [reply src/Stratis.SmartContracts.CLR/EcRecoverProvider.cs]: Fixed

**Pull request closed without merging**